### PR TITLE
mpir_pmi: allow runtime pmi selection

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description
Instead of selecting a PMI API version at configure time, detect all available PMI versions, and enable multiple APIs if they are supported. Use `MPIR_CVAR_PMI_VERSION` environment variable to select a runtime PMI protocol.

[skip warnings]
## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [ ] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/pmi

**Comment by hzhou** [comment]
test:mpich/pmi

**Comment by hzhou** [comment]
test:mpich/pmi

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

File: src/util/mpir_pmi.c [hunk 1]
```diff
@@ -730,9 +730,12 @@ static int optional_bcast_barrier(MPIR_PMI_DOMAIN domain)
         /* PMI2 local uses Put/GetNodeAttr, no need for barrier */
         return MPI_SUCCESS;
     }
-#elif defined(USE_PMIx_API)
-    /* PMIx will block/wait, so barrier unnecessary */
-    return MPI_SUCCESS;
+#elif defined(USE_PMIX_API)
+    if (domain == MPIR_PMI_DOMAIN_LOCAL) {
+        MPIR_pmi_barrier_local();
```
Comment (raffenet) [inline src/util/mpir_pmi.c]: so we do both local and full barrier in this case?

Comment (hzhou) [reply src/util/mpir_pmi.c]: Hmm, should be `return MPIR_pmi_barrier_local();`

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

File: src/util/mpir_pmi.c (hunk 1 above)
Comment (hzhou) [reply src/util/mpir_pmi.c]: Fixed.

**Review by raffenet — APPROVED** [review]

**Pull request merged**